Feature/block illumination

reckoncrafter opened the pull request:
WARNING! SIGNIFICANT REFACTOR!

# `ExecutionDirector`

The `ExecutionDirector` completely takes over the behavior of `QueueReading` and much of `TurtleMovement`, centralizing it such that there can be a association between the the functions called in `TurtleMovement` and the blocks that represent them. This, conveniently, makes it trivial to enable illumination of the blocks during execution, which is defined in the function `IEnumerator IlluminateBlock(GameObject block)`.

It has the following components:
- `InstructTurtle()` and `CombinedJump()` which resolves the `Command` enum to the associated `TurtleMovement` function. `CombinedJump()` handles the special case where a `Jump` needs to be combined with the action following it, should there be one. **NOTE**: This combining behavior currently does not cross control flow boundaries.
- `EvaluateCondition()`, which resolves the conditions in the `Command` enum to the associated `TurtleMovement` function, and returns the result.
- `ScopeData` is a `struct` used to keep track of control flow statements.
- `FunctionData` is a `struct` used to keep track of function calls.
- `TurtleSyntaxError` and `TurtleRuntimeError`, unique `Exception`s used to halt the `Turtle` if it detects something wrong with the block program.

### `AssembleScopes()`
is run before the main blocks are executed, and before `FunctionBlock`s are executed.
It is responsible for assembling `Dictionary<int, ScopeData> FunctionData.scopeDict`, which control flow statements use to know where to move the `instructionPointer` to.

It does this by using a stack: `scopeBuilding`, pushing new `ScopeData`s when a Begin block is encountered, and popping them when an End block is encountered. This allows for control flow statements to be nested.

This process is where any illegally placed control flow blocks are detected, such as `IfBegin`s with no `IfEnd`, `IfBegin`s ended with `WhileEnd`s, `Else` or `ElseIf`s with no `IfBegin`, etc. Such detections will throw `TurtleSyntaxError`.

### `MainLoop()`
is an event-driven function that handles the primary execution logic of the block program.

It is triggered by two different `UnityEvent`s: `ContinueLoopEvent` is `Invoke()`d whenever any block that does not directly cause the turtle to move is evaluated, and `TurtleMovement.EndMovement`, which means that `MainLoop()` will always wait for the `Turtle` to finish moving before executing the next block in the list. `MainLoop()` will unhook itself from these events whenever the turtle fails, succeeds, or the program is over. Pressing the `StartButton` rehooks it.

When the loop needs to retrigger itself with `ContinueLoopEvent` it calls `StartCoroutine(RepeatLoop())`, which currently enforces a 1 second delay between the next call of the loop. This allows for the user to visually observe the execution process, watching how the program handles if statements, while statements, and functions.

#### The `Stack<FunctionData> CallStack`

In order to handle functions, the loop needs to be able to switch to a new list of blocks, while keeping track of its position in the list that called the function. As such `FunctionData` holds references to all the crucial information about the current list:
- `List<GameObject> blockList`: the list itself, such as the list collected under the `StartQueue` block, or a `FunctionBlock`.
- `int instructionPointer`: the current position in the list, and thus where to resume from when a function is complete.
- `Dictionary<int, ScopeData> scopeDict`: the control flow statements associated with this list.
- `Stack<ScopeData> scopes`: for any control flow statements we are currently inside.

Whenever a function is called, a new `FunctionData` representing the function is pushed to the `CallStack` and the loop is called again, where it will now begin executing the function. Whenever the function is complete, it will `Pop()` the `CallStack`, and pickup exactly where it left off. This means that functions can be nested, and theoretically be recursive, though additional error handling is needed for this behavior.

# `TurtleMovement`
has been significantly altered. Most notable, it no longer contains a `Queue<Action>` and no longer controls its own behavior.

This was one of the trickiest parts of this refactor, as `TurtleMovement` did a great deal of coordination *between* each action, such as calling `PerformJump()` simultaneously with the action that proceeds it. I have tried to preserve this coordination as well as I can, but there may yet be some issues.

In order to properly control the behavior of the `ExecutionDirector`, `TurtleMovement` now has a series of `UnityEvent`s, which `ExecutionDirector` listens to.

## Major Changes
- Completely Removed `QueueReading`, made redundant by `ExecutionDirector`.
- `TurtleCommand` has significantly less responsibility. The only active behavior it has is to setup the `OffendingStateMaterial` so errors can be visually identified. Otherwise, it serves only to host the `TurtleCommand.Command` `enum`, which determines `ExecutionDirector`'s actions.
- Because the `Turtle` no longer contains a `Queue<Action>` the `MovementMenu` now contains **direct** instructions. The `Turtle` will move immediately when a menu button is pressed. The conditions are no longer in the menu. These can be added again for debugging purposes if desired.
- `StartButton` no longer has any inherent behavior. `ExecutionDirector` is configured to reference it in the Editor, and hooks into its `onSelectEntered()` event upon startup.
- Added `ExecutionDirector` to all levels.

## Minor Changes
- Removed `IncompleteConditionalHandler.cs`, made redundant by `ExecutionDirector` and changes to `TurtleCommand`
- Modified `BlockCount.cs` to fetch from `ExecutionDirector`
- Added `blockSnapEvent` to `BlockSnapping.cs`, which triggers `ExecutionDirector` to update its lists.
- `DeletionBoundary` no longer triggers the Turtle to `Reset()`, as it resets on its own.
- `Block (StartQueue)` no longer has a `QueueReading` component.
- `FunctionBlock` no longer has a `QueueReading` component.
- `FunctionCallBlock` no longer has any behavior besides referencing it's `FunctionBlock`.

malcolmcase97 reviewed (changes requested):
Blocked until CombinedJump bug is worked out! (See Discord/Pull-Requests)

reckoncrafter commented:
This latest commit fixes the bugs encountered when `CombinedJump()` is triggered. I hypothesized that the issue was caused by `CombinedJump()` triggering two movement commands simultaneously, and thus causing `turtleMovement.EndOfMovementEvent` to be `Invoke()`d twice, causing `MainLoop()` to trigger twice, and create completely bugged execution. These changes remove `MainLoop()`'s dependency on `EndOfMovementEvent`, and instead has every code path call `RepeatLoop()` at its end.

I modified `RepeatLoop()` to take a `float` argument, which will determine how long it should wait before triggering the loop again. Three `float` constants, `defaultWait` for control flow blocks, `moveWait` for the lateral movement blocks, and `jumpWait` for jumping and combined jumping, have been defined above `void MainLoop()`. I feel I've given them reasonable values here, but feel free to tune them.

I also changed `IlluminateBlock()` to also take a `time` parameter, making the illumination synchronized with the `RepeatLoop()` delay, and I've also included additional calls to this function to illuminate condition blocks at the same time as the block evaluating them, and movement blocks that are combined with Jump. This makes the block illumination much more comprehensible and visually satisfying.

I fixed `BlockCount` by having it listen to `BlockSnapping.blockSnapEvent` to update itself.
